File: libs/src/csharp/SalmonFS/FS/Streams/WSFileStream.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

using System;
using System.IO;
using System.Net.Http;
using System.Net;
using System.Text;
using System.Net.Http.Headers;
using Mku.Streams;
using System.Threading.Tasks;
using Mku.Convert;
using System.Web;
using System.Collections.Specialized;
using System.Runtime.CompilerServices;
using Mku.FS.File;
using static Mku.FS.File.HttpFile;
using Mku.Salmon.Encode;

namespace Mku.FS.Streams;


/// <summary>
/// A Web Service File Stream implementation.
/// This class is used internally for random file access of remote physical (real) files.
/// </summary>
public class WSFileStream : RandomAccessStream
{
    private static readonly string PATH = "path";
    private static readonly string POSITION = "position";
    private static readonly string LENGTH = "length";

    /// <summary>
    /// The HTTP client
    /// </summary>
    public static HttpSyncClient Client { get; set; } = HttpSyncClient.Instance;
    private Stream inputStream;
    private Stream outputStream;
    private bool closed;
    private WSFile file;

    /// <summary>
    /// Maximum amount of bytes allowed to skip forwards when seeking otherwise will open a new connection
    /// </summary>
    public long MaxNetBytesSkip { get; set; } = 32768;

    private bool canWrite;
    private long position;
    private HttpResponseMessage httpResponse;
    private HttpResponseMessage outHttpResponse;

    /// <summary>
    /// Stream for web service file.
    /// </summary>
    /// <param name="file">The web service file</param>
    /// <param name="access">The access mode</param>
    public WSFileStream(WSFile file, FileAccess access)
    {
        this.file = file;
        if (access == FileAccess.Write)
        {
            this.canWrite = true;
        }
    }

    /// <summary>
    /// Check if stream can read.
    /// </summary>
    public override bool CanRead => !canWrite;

    /// <summary>
    /// Check if stream can seek.
    /// </summary>
    public override bool CanSeek => true;

    /// <summary>
    /// Check if stream can write.
    /// </summary>
    public override bool CanWrite => canWrite;

    /// <summary>
    /// The length of the stream.
    /// </summary>
    public override long Length => file.Length;

    /// <summary>
    /// The position of the stream.
    /// </summary>
    public override long Position
    {
        get
        {
            return position;
        }
        set
        {
            // if the new position is forwards we can skip a small amount rather opening up a new connection
            if (this.position < value && value - position < MaxNetBytesSkip && this.inputStream != null)
            {
                byte[] buffer = new byte[32768];
                int totalBytesRead = 0;
                int bytesRead;
                while ((bytesRead = inputStream.Read(buffer, 0, Math.Min(buffer.Length, (int)((value - position) - totalBytesRead)))) > 0)
                {
                    totalBytesRead += bytesRead;
                }
            }
            else if (this.position != value)
            {
                this.Reset();
            }
            this.position = value;
        }
    }

    /// <summary>
    /// Flush the stream.
    /// </summary>
    public override void Flush()
    {
        try
        {
            if (outputStream != null)
                outputStream.Flush();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
        }
    }

    /// <summary>
    /// Read from the stream.
    /// </summary>
    /// <param name="buffer">The buffer</param>
    /// <param name="offset">The offset</param>
    /// <param name="count">The number of bytes to read</param>
    /// <returns></returns>
    public override int Read(byte[] buffer, int offset, int count)
    {
        int res = GetInputStream().Read(buffer, offset, count);
        position += res;
        return res;
    }

    /// <summary>
    /// Seek to a position.
    /// </summary>
    /// <param name="offset">The offset</param>
    /// <param name="origin">The type of seek</param>
    /// <returns></returns>
    public override long Seek(long offset, SeekOrigin origin)
    {
        long pos = this.Position;

        if (origin == SeekOrigin.Begin)
            pos = offset;
        else if (origin == SeekOrigin.Current)
            pos += offset;
        else if (origin == SeekOrigin.End)
            pos = file.Length - offset;

        this.Position = pos;
        return this.Position;
    }

    /// <summary>
    /// Set the length of the stream.
    /// </summary>
    /// <param name="value"></param>
    /// <exception cref="IOException"></exception>
    public override void SetLength(long value)
    {
        if (this.closed)
            throw new IOException("Stream is closed");
        HttpResponseMessage httpResponse = null;
        try
        {
            UriBuilder builder = new UriBuilder(file.ServicePath + "/api/setLength");
            NameValueCollection query = HttpUtility.ParseQueryString(builder.Query);
            query[PATH] = file.Path;
            query[LENGTH] = value + "";
            builder.Query = query.ToString();
            string url = builder.ToString();

            HttpRequestMessage requestMessage = new HttpRequestMessage(HttpMethod.Get, url);
            SetDefaultHeaders(requestMessage);
            SetServiceAuth(requestMessage);
            httpResponse = Client.Send(requestMessage);
            CheckStatus(httpResponse, HttpStatusCode.OK);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
            throw;
        }
        finally
        {
            if (httpResponse != null)
                httpResponse.Dispose();
        }
        Reset();
    }

    /// <summary>
    /// Write to the stream.
    /// </summary>
    /// <param name="buffer">The buffer</param>
    /// <param name="offset">The offset</param>
    /// <param name="count">The number of bytes to write</param>
    public override void Write(byte[] buffer, int offset, int count)
    {
        GetOutputStream().Write(buffer, offset, count);
        position += Math.Min(buffer.Length, count);
    }

	[MethodImpl(MethodImplOptions.Synchronized)]
    private Stream GetInputStream()
    {
        if (this.closed)
            throw new IOException("Stream is closed");
        if (this.inputStream == null)
        {
            long startPosition = this.Position;
            try
            {
                UriBuilder builder = new UriBuilder(file.ServicePath + "/api/get");
                NameValueCollection query = HttpUtility.ParseQueryString(builder.Query);
                query[PATH] = file.Path;
                query[POSITION] = startPosition + "";
                builder.Query = query.ToString();
                string url = builder.ToString();

                HttpRequestMessage requestMessage = new HttpRequestMessage(HttpMethod.Get, url);
                SetDefaultHeaders(requestMessage);
                SetServiceAuth(requestMessage);
                httpResponse = Client.Send(requestMessage);
                CheckStatus(httpResponse, startPosition > 0 ? HttpStatusCode.PartialContent : HttpStatusCode.OK);
                Stream stream = httpResponse.Content.ReadAsStream();
                this.inputStream = new BufferedStream(stream, DEFAULT_BUFFER_SIZE);
                return stream;
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine(ex);
                Close();
                throw;
            }
        }
        if (this.inputStream == null)
            throw new IOException("Could not retrieve stream");
        return this.inputStream;
    }

    private Stream GetOutputStream()
    {
        if (this.closed)
            throw new IOException("Stream is closed");
        if (this.outputStream == null)
        {
            HttpRequestMessage requestMessage = null;
            BlockingInputOutputAdapterStream outputStream = null;
            long startPosition = this.Position;
            try
            {
                outputStream = new BlockingInputOutputAdapterStream();
                this.outputStream = outputStream;

                UriBuilder builder = new UriBuilder(file.ServicePath + "/api/upload");
                NameValueCollection query = HttpUtility.ParseQueryString(builder.Query);
                query[PATH] = file.Path;
                query[POSITION] = startPosition + "";
                builder.Query = query.ToString();
                string url = builder.ToString();

                requestMessage = new HttpRequestMessage(HttpMethod.Post, url);
                MultipartFormDataContent content = new MultipartFormDataContent();
                Stream pipedInputStream = outputStream.InputStream;
                content.Add(new StreamContent(pipedInputStream), "file", file.Name);
                requestMessage.Content = content;
                SetDefaultHeaders(requestMessage);
                SetServiceAuth(requestMessage);

                Task.Run(() =>
                {
                    try
                    {
                        outHttpResponse = Client.Send(requestMessage);
                        CheckStatus(outHttpResponse, startPosition > 0 ? HttpStatusCode.PartialContent : HttpStatusCode.OK);
                        outputStream.SetReceived(true);
                    }
                    catch (Exception e)
                    {
                        Console.Error.WriteLine(e);
                        throw;
                    }
                    finally
                    {
                        if (this.outputStream != null)
                        {
                            try
                            {
                                this.outputStream.Close();
                            }
                            catch (IOException e)
                            {
                                Console.Error.WriteLine(e);
                                throw;
                            }
                        }
                        if (outHttpResponse != null)
                        {
                            try
                            {
                                outHttpResponse.Dispose();
                            }
                            catch (IOException e)
                            {
                                Console.Error.WriteLine(e);
                                throw;
                            }
                        }
                    }
                });
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e);
                throw;
            }
        }
        if (this.outputStream == null)
            throw new IOException("Could not retrieve stream");
        return this.outputStream;
    }

    /// <summary>
    /// Close the stream.
    /// </summary>
    public override void Close()
    {
        Reset();
        this.closed = true;
    }

    private void Reset()
    {
        if (this.inputStream != null)
            this.inputStream.Close();
        this.inputStream = null;

        if (this.outputStream != null)
            this.outputStream.Close();
        this.outputStream = null;

        if (httpResponse != null)
            httpResponse.Dispose();
        if (outHttpResponse != null)
            outHttpResponse.Dispose();

        file.Reset();
    }

    private void CheckStatus(HttpResponseMessage httpResponse, HttpStatusCode status)
    {
        if (httpResponse.StatusCode != status)
        {
            throw new IOException(httpResponse.StatusCode
                    + " " + httpResponse.ReasonPhrase);
        }
    }

    private void SetServiceAuth(HttpRequestMessage httpRequestMessage)
    {
        if (this.file.ServiceCredentials != null)
        {
            string encoding = Base64Utils.Base64.Encode(UTF8Encoding.UTF8.GetBytes(file.ServiceCredentials.ServiceUser + ":" + file.ServiceCredentials.ServicePassword));
            httpRequestMessage.Headers.Authorization = new AuthenticationHeaderValue("Basic", encoding);
        }
    }

    private void SetDefaultHeaders(HttpRequestMessage requestMessage)
    {
        requestMessage.Headers.Add("Cache", "no-store");
    }
}

File: libs/src/python/salmon_fs/salmon_fs/fs/streams/ws_file_stream.py
#!/usr/bin/env python3
"""!@brief Stream implemetation for web service files.
"""

__license__ = """
MIT License

Copyright (c) 2025 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from typeguard import typechecked
from queue import Queue
from threading import Thread
from http.client import HTTPResponse, HTTPConnection, HTTPSConnection
import urllib
from urllib import parse
from io import RawIOBase

from salmon_core.convert.base_64 import Base64
from salmon_fs.fs.file.ifile import IFile
from salmon_fs.fs.file.http_sync_client import HttpSyncClient
from salmon_core.streams.random_access_stream import RandomAccessStream


@typechecked
class WSFileStream(RandomAccessStream):
    """!
    Stream implemetation for web service files.
    This class is used internally for random file access of physical (real) files.
    """
    __PATH: str = "path"
    __POSITION: str = "position"
    __LENGTH: str = "length"

    def __init__(self, file: IFile, mode: str):
        """!
        Construct a file stream from an AndroidFile.
        This will create a wrapper stream that will route read() and write() to the file

        @param file: The file that will be used to get the read/write stream
        @param mode: The mode "r" for read "rw" for write
        """

        self.__file: IFile | None = file
        """
        The python file associated with this stream.
        """

        self.position: int = 0

        self.__canWrite: bool = mode == "rw"
        self.__response: HTTPResponse | None = None
        self.conn: HTTPConnection | HTTPSConnection | None = None
        self.queue: Queue | None = None
        self.start_position: int = 0
        self.upload_thread: Thread | None = None
        self.closed: bool = False

    def __get_input_response(self) -> HTTPResponse:
        if self.closed:
            raise IOError("Stream is closed")
        if not self.__response:
            headers = {}
            self.__set_default_headers(headers)
            self.__set_service_auth(headers)
            params = urllib.parse.urlencode({WSFileStream.__PATH: self.__file.get_path(),
                                             WSFileStream.__POSITION: self.position})
            self.conn = self.__create_connection(self.__file.get_service_path())
            self.conn.request("GET", "/api/get" + "?" + params, headers=headers)
            self.__response = self.conn.getresponse()
            self.__check_status(self.__response, 206 if self.position > 0 else 200)
        return self.__response

    def get_output_queue(self) -> Queue:
        if self.closed:
            raise IOError("Stream is closed")
        if not self.queue:
            self.start_position: int = self.get_position()
            boundary = "*******"
            header = "--" + boundary + "\r\n"
            header += "Content-Disposition: form-data; name=\"file\"; filename=\"" \
                      + self.__file.get_name() + "\"\r\n"
            header += "\r\n"
            header_data = bytearray(header.encode())

            footer = "\r\n--" + boundary + "--"
            footer_data: bytearray = bytearray(footer.encode())

            @typechecked
            class PipedIO(RawIOBase):
                def __init__(self, queue: Queue):
                    self.is_closed = False
                    self.buff: bytearray | None = None
                    self.queue = queue

                def read(self, size: int | None = ...) -> bytearray | None:
                    if self.is_closed:
                        return None
                    if self.buff and len(self.buff) > 0:
                        buf = self.buff[0:size]
                        self.buff = self.buff[size:]
                        return buf
                    buffer: bytearray = bytearray()
                    while message := self.queue.get():
                        buffer.extend(message)
                        if len(buffer) > size:
                            self.buff = buffer[size:]
                            buffer = buffer[0:size]
                            break
                    if message is None:
                        buffer.extend(footer_data)
                        self.is_closed = True
                    return buffer

            self.queue = Queue()
            self.queue.put(header_data)
            data = PipedIO(self.queue)

            headers = {}
            self.__set_default_headers(headers)
            headers["Content-type"] = "multipart/form-data;boundary=" + boundary
            self.__set_service_auth(headers)
            params = urllib.parse.urlencode({WSFileStream.__PATH: self.__file.get_path(),
                                             WSFileStream.__POSITION: self.start_position})
            self.conn = self.__create_connection(self.__file.get_service_path())

            def start_upload():
                self.conn.request("POST", "/api/upload" + "?" + params, headers=headers, body=data)
                self.__response = self.conn.getresponse()

            self.upload_thread = Thread(target=start_upload)
            self.upload_thread.start()

        return self.queue

    def can_read(self) -> bool:
        """!
        True if stream can read from file.
        @returns True if readable
        """
        return not self.__canWrite

    def can_write(self) -> bool:
        """!
        True if stream can write to file.
        @returns True if writable
        """
        return self.__canWrite

    def can_seek(self) -> bool:
        """!
        True if stream can seek.
        @returns True if seekable
        """
        return True

    def get_length(self) -> int:
        """!
        Get the length of the stream. This is the same as the backed file.
        @returns The length
        """
        return self.__file.get_length()

    def get_position(self) -> int:
        """!
        Get the position of the stream
        @returns The position
        """
        return self.position

    def set_position(self, value: int):
        """!
        Set the current position of the stream.
        @param value: The new position.
        @exception IOError: Thrown if there is an IO error.
        """
        if self.position != value:
            self.reset()
        self.position = value

    def set_length(self, value: int):
        """!
        Set the length of the stream
        """
        headers = {}
        self.__set_default_headers(headers)
        self.__set_service_auth(headers)
        params = urllib.parse.urlencode({WSFileStream.__PATH: self.__file.get_path(),
                                         WSFileStream.__LENGTH: str(value)})
        conn: HTTPConnection | HTTPSConnection | None = None
        http_response: HTTPResponse | None = None
        try:
            conn = self.__create_connection(self.__file.get_service_path())
            conn.request("PUT", "/api/setLength", headers=headers, body=params)
            http_response = conn.getresponse()
            self.__check_status(http_response, 200)
            return True
        finally:
            if conn:
                conn.close()
            if http_response:
                http_response.close()

    def read(self, buffer: bytearray, offset: int, count: int) -> int:
        """!
        Read data from the file stream into the buffer provided.
        @param buffer: The buffer to write the data.
        @param offset: The offset of the buffer to start writing the data.
        @param count: The maximum number of bytes to read from.
        @returns The bytes read
        @exception IOError: Thrown if there is an IO error.
        """
        bytes_read: int = 0
        while bytes_read < count:
            buff: bytes = self.__get_input_response().read(count - bytes_read)
            if not buff or len(buff) == 0:
                break
            buffer[offset + bytes_read:offset + bytes_read + len(buff)] = buff[:]
            bytes_read += len(buff)
        if bytes_read <= 0:
            return -1
        self.position += bytes_read
        return bytes_read

    def write(self, buffer: bytearray, offset: int, count: int):
        """!
        Write the data from the buffer provided into the stream.
        @param buffer: The buffer to read the data from.
        @param offset: The offset of the buffer to start reading the data.
        @param count: The maximum number of bytes to read from the buffer.
        @exception IOError: Thrown if there is an IO error.
        """
        buff = buffer[offset:offset + count]
        queue = self.get_output_queue()
        queue.put(buff)
        self.position += len(buff)

    def seek(self, offset: int, origin: RandomAccessStream.SeekOrigin) -> int:
        """!
        Seek to the offset provided.
        @param offset: The position to seek to.
        @param origin: The type of origin {@link RandomAccessStream.SeekOrigin}
        @returns The new position after seeking.
        @exception IOError: Thrown if there is an IO error.
        """
        pos: int = self.position
        if origin == RandomAccessStream.SeekOrigin.Begin:
            pos = offset
        elif origin == RandomAccessStream.SeekOrigin.Current:
            pos += offset
        elif origin == RandomAccessStream.SeekOrigin.End:
            pos = self.__file.get_length() - offset

        self.set_position(pos)
        return self.position

    def flush(self):
        """!
        Flush the buffers to the associated file.
        """
        # TODO: flush network upload before queueing more?
        pass

    def close(self):
        """!
        Close this stream and associated resources.
        @exception IOError: Thrown if there is an IO error.
        """
        self.reset()
        self.closed = True

    def reset(self):
        """!
        Reset the stream
        """
        if self.can_write():
            if self.queue:
                self.queue.put(None)
            if self.upload_thread:
                self.upload_thread.join()
            if self.__response:
                self.__check_status(self.__response, 206 if self.start_position > 0 else 200)
                self.__response.close()
            if self.conn:
                self.conn.close()
        else:
            if self.__response:
                self.__response.close()
            if self.conn:
                self.conn.close()

        self.conn = None
        self.__response = None
        self.queue = None
        self.upload_thread = None
        self.start_position = 0
        self.__file.reset()

    def __set_service_auth(self, headers: dict[str, str]):
        if not self.__file.get_credentials():
            return
        headers['Authorization'] = 'Basic ' + Base64().encode(
            bytearray((self.__file.get_credentials().get_service_user() + ":"
                       + self.__file.get_credentials().get_service_password())
                      .encode('utf-8')))

    def __check_status(self, http_response: HTTPResponse, status: int):
        if http_response.status != status:
            raise IOError(str(http_response.status)
                          + " " + http_response.reason)

    def __set_default_headers(self, headers: dict[str, str]):
        headers["Cache"] = "no-store"
        headers["Content-type"] = "application/x-www-form-urlencoded"

    def __create_connection(self, url: str) -> HTTPConnection:
        conn: HTTPConnection = HttpSyncClient.create_connection(url)
        return conn
